### PR TITLE
[Direct Message] Enhance invite by email

### PR DESCRIPTION
**Pull request by Florian14**
#247 

Improve discovery results from DM creation screen and move logical from the invite by email of the home screen in a dedicated ViewModel

### CODE REVIEW

File: vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            return GetPlatformResult.Failure(Failure.ServerError(MatrixError(
+                    code = MatrixError.M_UNKNOWN,
+                    message = "No host"
+            ), 403))
+        }
+
+        var failure: Throwable? = null
+        val platform = idServerUrls.firstNotNullOfOrNull { url ->
+            try {
+                matrix.threePidPlatformDiscoverService().getPlatform(url, ThreePid.Email(params.email))
```
Comment (yostyle) [inline vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt:66]: Please restore success log

File: vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            } catch (throwable: Throwable) {
+                failure = throwable
```
Comment (yostyle) [inline vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt:68]: Please restore failure log

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomViewEvents.kt [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 New Vector Ltd
+ * Copyright (c) 2021 New Vector Ltd
```
Comment (yostyle) [inline vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomViewEvents.kt:2]: Please undo this change

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomViewModel.kt [hunk 4]
```diff
@@ -52,56 +56,136 @@ class CreateDirectRoomViewModel @AssistedInject constructor(@Assisted
...
      */
     private fun onSubmitInvitees(action: CreateDirectRoomAction.CreateRoomAndInviteSelectedUsers) {
-        val existingRoomId = action.selections.singleOrNull()?.getMxId()?.let { userId ->
-            session.getExistingDirectRoomWithUser(userId)
-        }
-        if (existingRoomId != null) {
-            // Do not create a new DM, just tell that the creation is successful by passing the existing roomId
-            setState {
-                copy(createAndInviteState = Success(existingRoomId))
+        val selection = action.selections.singleOrNull() ?: return
```
Comment (yostyle) [inline vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomViewModel.kt:68]: Please add some description to explain this trick for tchap

**Review by yostyle — APPROVED** [review]

File: vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        private val context: Context,
+        private val activeSessionHolder: ActiveSessionHolder,
+        private val matrix: Matrix
+) : ViewModelTask<Params, GetPlatformResult> {
+
+    override suspend fun execute(params: Params): GetPlatformResult {
+        Timber.d("## TchapGetPlatformTask [${params.email}]")
+
+        // Prepare the list of the known IS urls in order to run over the list until to get an answer.
+        val idServerUrls = buildIdServerUrls().toMutableList()
```
Comment (ClaireGizard) [inline vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt:55]: you don't need to use "toMutableList()"

File: vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            idServerUrls.add(currentIdServerUrl)
+        }
+
+        // Add randomly the preferred known ISes
+        var currentHosts: MutableList<String> = mutableListOf(*context.resources.getStringArray(R.array.preferred_identity_server_names))
+        while (currentHosts.isNotEmpty()) {
+            val index = Random().nextInt(currentHosts.size)
+            val host: String = currentHosts.removeAt(index)
+            val idServerUrl = context.getString(R.string.server_url_prefix) + host
+            if (null == currentIdServerUrl || idServerUrl != currentIdServerUrl) {
```
Comment (ClaireGizard) [inline vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt:94]: i prefer : "currentIdServerUrl == null"

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomViewModel.kt [hunk 7]
```diff
@@ -52,56 +56,151 @@ class CreateDirectRoomViewModel @AssistedInject constructor(@Assisted
...
+                        _viewEvents.post(CreateDirectRoomViewEvents.Failure(it))
                     }
+            )
+        }
+    }
 
-            val result = runCatchingToAsync {
-                session.createRoom(roomParams)
+    private fun handleIndividualInviteByEmail(email: String) {
+        setState { copy(isLoading = true) }
```
Comment (ClaireGizard) [inline vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomViewModel.kt:107]: i think you don't need this line here, you already have an : " setState { copy(isLoading = true) }" line 73.

File: vector/src/main/res/values-fr/strings.xml [hunk 8]
```diff
@@ -2964,7 +2964,7 @@
     <string name="you_will_leave_all_in">Vous allez quitter tous les salons et espaces de %s.</string>
     <string name="leave_all_rooms_and_spaces">Quitter tous les salons et espaces</string>
     <string name="space_leave_prompt_msg_with_name">Voulez-vous vraiment quitter %s \?</string>
-    <string name="discovery_section">Découverte (%s)</string>
+    <string name="discovery_section">Utilisateur découvert</string>
```
Comment (ClaireGizard) [inline vector/src/main/res/values-fr/strings.xml:2967]: Maybe we can use: "Utilisateur trouvé", i think it's weird in french.

File: vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt (hunk 6 above)
Comment (Florian14) [reply vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt:94]: Me too, I moved this code from the PlatformViewModel and this is a convention that we find in many other places in the project

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomViewModel.kt (hunk 7 above)
Comment (Florian14) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomViewModel.kt:107]: This method is also called line 59, in this case the state will no be updated if I remove it

File: vector/src/main/res/values-fr/strings.xml (hunk 8 above)
Comment (Florian14) [reply vector/src/main/res/values-fr/strings.xml:2967]: We should discuss about that with @giomfo

File: vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt (hunk 5 above)
Comment (Florian14) [reply vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt:55]: well seen

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomViewModel.kt (hunk 7 above)
Comment (ClaireGizard) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomViewModel.kt:107]: Maybe you can move: "    setState { copy(isLoading = true) }" line 57, then you can delete it in: onSubmitInvitees, handleIndividualInviteByEmail and handleCreateDirectMessageByUserId

Comment (Florian14) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomViewModel.kt:107]: I prefer to manage it separately.
Suppose we add an action which does not require a loading state, we'll be forced to deal with it. Also, at the beggining of onSubmitInvite, I don't update the state if the selection is bad, so I'll have to deal with it again

File: vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt (hunk 6 above)
Comment (yostyle) [reply vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt:94]: You can ask @bmarty about this.

Comment (bmarty) [reply vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt:94]: 👋 
Using `null == currentIdServerUrl` is an old Java convention, to avoid mistaking between:
`currentIdServerUrl == null` and
`currentIdServerUrl = null` which can lead to hard-to-see bugs.

I think in Kotlin there is no reason not to use `currentIdServerUrl == null`, since you cannot write 
```kotlin
if(currentIdServerUrl = null) {...}
```
or you will get the error `Assignments are not expressions, and only expressions are allowed in this context`

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomActivity.kt [hunk 9]
```diff
@@ -171,14 +189,21 @@ class CreateDirectRoomActivity : SimpleFragmentActivity() {
         }
     }
 
-    private fun renderCreationSuccess(roomId: String?) {
+    private fun renderCreationSuccess(roomId: String) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomActivity.kt:192]: It will be hard to report every little fixes like that in Element but ideally this is what you should do (or fix only on Element and get it back during the next merge).

Comment (Florian14) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomActivity.kt:192]: Done: https://github.com/vector-im/element-android/pull/4757/commits/c13feb983ec6b05437aa5cb76fc3a7c35cad52ba

File: vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt (hunk 6 above)
Comment (Florian14) [reply vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt]: Fixed

File: vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt [hunk 10]
```diff
@@ -76,37 +74,22 @@ class TchapGetPlatformTask @Inject constructor(
...
-        while (currentHosts.isNotEmpty()) {
-            val index = Random().nextInt(currentHosts.size)
-            val host: String = currentHosts.removeAt(index)
-            val idServerUrl = context.getString(R.string.server_url_prefix) + host
-            if (currentIdServerUrl == null || idServerUrl != currentIdServerUrl) {
-                idServerUrls.add(idServerUrl)
-            }
-        }
+        // Compute list randomly from the preferred known ISes
+        val knownHosts = context.resources.getStringArray(R.array.preferred_identity_server_names).toList().shuffled()
```
Comment (yostyle) [inline vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt:86]: I like your suggestion. Just one remark. Could you replace toList() by toSet() ?

Comment (yostyle) [inline vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt:86]: I like your suggestion. Just one remark. Could you replace toList() by toSet() ?

File: vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt [hunk 11]
```diff
@@ -76,37 +74,22 @@ class TchapGetPlatformTask @Inject constructor(
...
+                // add randomly the other known ISes
+                .plus(context.resources.getStringArray(R.array.identity_server_names).toList().shuffled())
+                // map server names to server urls
+                .map { host -> context.getString(R.string.server_url_prefix) + host }
+        idServerUrls.addAll(knownHosts)
 
-        return idServerUrls
+        return idServerUrls.distinct()
```
Comment (yostyle) [inline vector/src/main/java/fr/gouv/tchap/features/platform/TchapGetPlatformTask.kt:93]: If you can replace the urls list by a set, also you should undo this change.

**Review by yostyle — APPROVED** [review]

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomActivity.kt (hunk 9 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomActivity.kt:192]: Thanks!
